proposals/ref-readonly-parameters.md

```diff
@@ -0,0 +1,112 @@
+- `ref readonly` → `in` would not be a breaking change (but `ref` callsite annotation would result in a warning).
+
+## Detailed design
+[design]: #detailed-design
+
+No changes in grammar are necessary.
+Specification will be extended to allow `ref readonly` modifiers for parameters with the same rules as specified for `in` parameters in [their proposal](https://github.com/dotnet/csharplang/blob/c8c1615fcad4ca016a97b7260ad497aad53ebc78/proposals/csharp-7.2/readonly-ref.md), except where explicitly changed in this proposal.
+
+### Interchangeability with `in`
+
+Members declared in a single type cannot differ in signature solely by `ref`/`out`/`in`/`ref readonly`.
+For other purposes of signature matching (e.g., hiding or overriding), `ref readonly` can be interchanged with `in` modifier, but that results in a warning at the declaration site [[§7.6](https://github.com/dotnet/csharpstandard/blob/47912d4fdae2bb8c3750e6485bdc6509560ec6bf/standard/basic-concepts.md#76-signatures-and-overloading)].
+For the purpose of delegate/lambda/method group conversions [[§10.7](https://github.com/dotnet/csharpstandard/blob/47912d4fdae2bb8c3750e6485bdc6509560ec6bf/standard/conversions.md#107-anonymous-function-conversions), [§10.8](https://github.com/dotnet/csharpstandard/blob/47912d4fdae2bb8c3750e6485bdc6509560ec6bf/standard/conversions.md#108-method-group-conversions)], `ref readonly` can be interchanged with `in` modifier, but that results in a warning.
+
+### Metadata encoding
+[metadata]: #metadata-encoding
+
+As a reminder,
+- `ref` parameters are emitted as plain byref types (`T&` in IL),
+- `in` parameters are like `ref` plus they are annotated with `System.Runtime.CompilerServices.IsReadOnlyAttribute`.
+  In C# 7.3 and later, they are also emitted with `[in]` and if virtual, `modreq(System.Runtime.InteropServices.InAttribute)`.
+
+`ref readonly` parameters will be emitted as `[in] T&`, plus annotated with the following attribute:
+
+```cs
+namespace System.Runtime.CompilerServices
+{
+    [AttributeUsage(AttributeTargets.Parameter, AllowMultiple = false, Inherited = false)]
+    public sealed class RequiresLocationAttribute : Attribute
+    {
+    }
+}
+```
+
+Furthermore, if virtual, they will be emitted with `modreq(System.Runtime.InteropServices.InAttribute)` to ensure binary compatibility with `in` parameters.
+Note that unlike `in` parameters, no `[IsReadOnly]` is emitted for `ref readonly` parameters to avoid increasing metadata size.
+
+The `RequiresLocationAttribute` will be matched by namespace-qualified name and synthesized by the compiler if not already included in the compilation.
+
+#### Function pointers
+
+In function pointers, `in` parameters are emitted with `modreq(System.Runtime.InteropServices.InAttribute)`.
+`ref readonly` parameters will be emitted in the same way to ensure binary compatibility.
```

Comment:
Do we really want this behaviour?
Consider these 2 different options:

1. The one you have here:
This would mean that if I had it in a declaration that was used by another library, it would have to be treated as `in` anyway. A specific example:
```C#
//Library1
public class Class1
{
  public static delegate*<in int, ref readonly int, void> Func1();
}
```
This would be emitted such that it looks like `delegate*<in int, in int, void>`, therefore in Library 2 we would be able to pass an rvalue with no warning to the second parameter.

2. Consider encoding `ref readonly` with an additional `modreq`. This would allow the calling library to see the correct parameter type on the above example. It does mean that changing from `in` to `ref readonly` in a function pointer type is a breaking change, but the alternative is that it does nothing and gets thrown away. I personally think a breaking change is more useful than throwing away the information altogether (and by useful, I mean I wouldn't change it for existing APIs, but at least for new APIs it would provide some value).
With this solution, you would want to only warn on code like
```csharp
//method declared somewhere
static void X(ref readonly int i) => ...;

//code
delegate*<in int, void> abc = &X; //warning
delegate*<ref readonly int, void> def = abc; //warning?
delegate*<ref int, void> = abc; //error without explicit casting, just like currently
```
because otherwise it would be a breaking change to change any method.

Benefits of (1): allows people to change from `in` to `ref readonly` without any breaking changes.
Drawbacks of (1): doesn't provide much value, since the information will be lost by any external library looking at it.

The benefits of (2): having `ref readonly` in as a parameter of a function pointer has a meaning to external libraries (otherwise this information is lost and they can pass an rvalue without a warning).
Drawbacks of (2): have to implement implicit casting with warning from something like `delegate*<ref readonly int, void>` to `delegage*<in int, void>` and vice versa to avoid any breaking changes. Potential modreq issues if we don't use a type that has existed for some period of time.

There is technically a third option, which is to emit complicated attributes that the compiler can use to determine which ones are `ref readonly` vs `in`.

I think if we don't go with (2), then there's no point allowing `ref readonly` parameters in function pointers at all, since they would be exactly the same as `in` parameters; meaning we wouldn't even be able to expose this information in the future without complex attribute tricks or breaking changes.

Also note with (2), the modreq would ideally be something that has existed previously, to avoid issues if were were to make the type declared in the runtime (since that would make it use a different assembly), or it could emit a type forward whenever it doesn't declare its own copy of the attribute, or it could leave it to me to fix in my own code 😆.

Reply:
Reading through the rest of the thing, (2) also works well with the `Interchangeability with in` section, since it would also allow for warnings when implicitly casting between something like `delegate*<ref readonly int, void>` and `delegage*<in int, void>`.

Specifically this part:
> For the purpose of delegate/lambda/method group conversions [[§10.7](https://github.com/dotnet/csharpstandard/blob/47912d4fdae2bb8c3750e6485bdc6509560ec6bf/standard/conversions.md#107-anonymous-function-conversions), [§10.8](https://github.com/dotnet/csharpstandard/blob/47912d4fdae2bb8c3750e6485bdc6509560ec6bf/standard/conversions.md#108-method-group-conversions)], `ref readonly` can be interchanged with `in` modifier, but that results in a warning.

And honestly, I'm thinking (3) (which is to use a complicated (to me anyway) attribute) would be a better solution than (2), but I have no idea how that would be encoded. Presumably, it could be encoded similarly to how names for value tuples are encoded, since they also have to encode things in relatively arbitrary locations.

Reply:
> in Library 2 we would be able to pass an rvalue with no warning to the second parameter.

Good point. I think we could emit a `modopt` for `ref readonly` parameters. It wouldn't be an error like emitting `modreq`, but could be used to emit warnings.